### PR TITLE
Not manage keys creation

### PR DESCRIPTION
**Pull request by rhoml**
We are running multiple puppetmasters and we would like to distribute eyaml keys using our own puppet repository for the public key and some other method for the private key. With this PR we can ensure this modules manage file permission on keys and directories.

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -118,7 +126,7 @@
     owner   => $keys_dir_owner,
     group   => $keys_dir_group,
     mode    => $public_key_mode,
-    require => Exec['eyaml_createkeys'],
+    require => $requires,
```
Comment (ghoneycutt) [inline manifests/init.pp]: Remove the require's on the files and use a `before` on the exec

File: manifests/init.pp [hunk 2]
```diff
@@ -31,6 +31,7 @@
   },
   $createkeys_path     = '/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin',
   $manage_eyaml_config = true,
+  $manage_keys         = true,
```
Comment (ghoneycutt) [inline manifests/init.pp]: This needs to be documented in the README.md

File: manifests/init.pp [hunk 3]
```diff
@@ -105,11 +106,18 @@
     require => Package['eyaml'],
   }
 
-  exec { 'eyaml_createkeys':
-    path    => $createkeys_path,
-    command => "eyaml createkeys --pkcs7-private-key=${private_key_path} --pkcs7-public-key=${public_key_path}",
-    creates => $private_key_path,
-    require => File['eyaml_keys_dir'],
+  if $manage_keys {
```
Comment (ghoneycutt) [inline manifests/init.pp]: `if manage_keys == true {`

File: manifests/init.pp [hunk 4]
```diff
@@ -105,11 +106,18 @@
...
+    exec { 'eyaml_createkeys':
+      path    => $createkeys_path,
+      command => "eyaml createkeys --pkcs7-private-key=${private_key_path} --pkcs7-public-key=${public_key_path}",
+      creates => $private_key_path,
+      require => File['eyaml_keys_dir'],
+    }
+  }
+
+  case $manage_keys {
```
Comment (ghoneycutt) [inline manifests/init.pp]: This conditional could be removed by using `before`

File: manifests/init.pp (hunk 2 above)
Comment (ghoneycutt) [inline manifests/init.pp]: We need to validate our inputs and when dealing with booleans, handle the case of stringified booleans. This can be done with

``` puppet
if is_string($manage_keys) == true {
  $manage_keys_real = str2bool($manage_keys)
} else {
  $manage_keys_real = $manage_keys
}
validate_bool($manage_keys_real)
```

**Comment by ghoneycutt** [comment]
Thanks @rhoml for your work, this is a great idea! Could you please make the noted changes and I will be happy to help with the spec tests, unless you would like to take a stab at those.

File: manifests/init.pp (hunk 2 above)
Comment (rhoml) [inline manifests/init.pp]: of we can just enforce it to be a bool right? I rather prefer people using the right values than giving them options to use strings

Comment (ghoneycutt) [inline manifests/init.pp]: We do enforce it with the code above. This is done because of Puppet's history of handling boolean values poorly. It aligns with the principal of least surprise, so if someone does pass a stringified boolean, things still work as expected.

File: manifests/init.pp (hunk 3 above)
Comment (rhoml) [inline manifests/init.pp]: This is not required since `if $manage_keys {`does this validation if we ensure it is a bool.

**Comment by rhoml** [comment]
Hey @ghoneycutt I've did a couple of things
- Added variable description with expected rtype on the init.pp
- Added some tests 
- Added pre-commit.com file useful for development
- Updated gitignore so we can use bundler
- Added the new minor version upgrade to the metadata.json file.

**Comment by rhoml** [comment]
@ghoneycutt any news on this?

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+-   repo: git://github.com/pre-commit/mirrors-puppet-lint
```
Comment (ghoneycutt) [inline .pre-commit-config.yaml]: what is this file?

File: metadata.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ghoneycutt-eyaml",
-  "version": "0.3.0",
+  "version": "0.3.1",
```
Comment (ghoneycutt) [inline metadata.json]: The maintainer will update this upon release. For instance, I might integrate other work into the release or merge ofter PR's first and release them, so you would not want to make these types of changes for version numbers.

File: manifests/init.pp [hunk 7]
```diff
@@ -21,7 +128,7 @@
   $config_dir_group    = 'root',
   $config_dir_mode     = '0755',
   $config_ensure       = 'file',
-  $config_path         = '/etc/eyaml/config.yaml',
+  $config_path         = "${::settings::confdir}/config.yaml",
```
Comment (ghoneycutt) [inline manifests/init.pp]: `settings` is a class.. this line should stay as it is, as you can already override with Hiera or a parameterized class

File: spec/classes/init_spec.rb [hunk 8]
```diff
@@ -1,7 +1,78 @@
 require 'spec_helper'
 describe 'eyaml' do
 
-  context 'with defaults for all parameters' do
-    it { should contain_class('eyaml') }
+describe 'eyaml', :type => :class do
```
Comment (ghoneycutt) [inline spec/classes/init_spec.rb:4]: Thanks for spec tests!!

File: manifests/init.pp [hunk 9]
```diff
@@ -2,6 +2,113 @@
 #
 # Module to manage eyaml
 #
+# Variables
+# ----------
+#
+# Here you should define a list of variables that this module would require.
+#
+# * `package_name`: hiera-eyaml requires the hiera-eyaml gem to be installed.
+#    rtype: string
```
Comment (ghoneycutt) [inline manifests/init.pp]: why `rtype` instead of `type` or `types` ?

File: manifests/init.pp [hunk 10]
```diff
@@ -2,6 +2,113 @@
...
+#
+# * `keys_dir_group`: Ensures that $keys_dir group is correct.
+#    rtype: string
+#    Default: $::settings::group
+#
+# * `keys_dir_mode`: Ensures the $keys_dir mode is -r-x------.
+#    rtype: integer
+#    Default: 0500
+#
+# * `public_key_path`: Absolute path to the public key.
```
Comment (ghoneycutt) [inline manifests/init.pp]: thanks for documenting all these

File: Rakefile [hunk 11]
```diff
@@ -1,18 +1,10 @@
 require 'puppetlabs_spec_helper/rake_tasks'
 require 'puppet-lint/tasks/puppet-lint'
+
+PuppetLint.configuration.fail_on_warnings = true
+PuppetLint.configuration.send('relative')
 PuppetLint.configuration.send('disable_80chars')
-PuppetLint.configuration.relative = true
```
Comment (ghoneycutt) [inline Rakefile]: this needs to be here

File: Rakefile [hunk 12]
```diff
@@ -1,18 +1,10 @@
...
+PuppetLint.configuration.send('disable_class_inherits_from_params_class')
```
Comment (ghoneycutt) [inline Rakefile]: these disables should be removed

File: Rakefile [hunk 13]
```diff
@@ -1,18 +1,10 @@
...
+PuppetLint.configuration.send('disable_documentation')
+PuppetLint.configuration.send('disable_single_quote_string_with_variables')
 PuppetLint.configuration.ignore_paths = ["spec/**/*.pp", "pkg/**/*.pp"]
-
-desc 'Validate manifests, templates, and ruby files'
```
Comment (ghoneycutt) [inline Rakefile:7]: why was this removed?

File: .sync.yml [hunk 14]
```diff
@@ -0,0 +1,9 @@
+---
```
Comment (ghoneycutt) [inline .sync.yml]: what is this file for?

File: Gemfile [hunk 15]
```diff
@@ -1,15 +1,15 @@
 source ENV['GEM_SOURCE'] || 'https://rubygems.org'
 
-if puppetversion = ENV['PUPPET_GEM_VERSION']
```
Comment (ghoneycutt) [inline Gemfile:3]: This needs to be here so we can do matrix testing of different puppet versions

File: Gemfile [hunk 16]
```diff
@@ -1,15 +1,15 @@
...
-  gem 'puppet', puppetversion, :require => false
-else
-  gem 'puppet', :require => false
-end
 
+gem 'rake'
```
Comment (ghoneycutt) [inline Gemfile]: these don't need to be here as they are just dependencies

File: Gemfile [hunk 17]
```diff
@@ -1,15 +1,15 @@
...
+gem 'mocha'
+gem 'diff-lcs'
+gem 'json_pure'
+gem 'json'
 gem 'metadata-json-lint'
-gem 'puppetlabs_spec_helper', '>= 0.1.0'
-gem 'facter', '>= 1.7.0'
-gem 'rspec-puppet'
+gem 'rspec-puppet', '2.2.0'
```
Comment (ghoneycutt) [inline Gemfile]: why pin at that exact version? we want the newest

File: Gemfile [hunk 18]
```diff
@@ -1,15 +1,15 @@
...
+gem 'puppet', '3.8.4'
```
Comment (ghoneycutt) [inline Gemfile]: See comment above regarding matrix testing

File: Gemfile [hunk 19]
```diff
@@ -1,15 +1,15 @@
...
+gem 'puppetlabs_spec_helper', '0.10.3'
```
Comment (ghoneycutt) [inline Gemfile]: again, we want the newest here

File: manifests/init.pp [hunk 20]
```diff
@@ -21,7 +128,7 @@
   $config_dir_group    = 'root',
   $config_dir_mode     = '0755',
   $config_ensure       = 'file',
-  $config_path         = '/etc/eyaml/config.yaml',
```
Comment (ghoneycutt) [inline manifests/init.pp:24]: Why change the location of the eyaml config? I believe it should be in /etc/eyaml and not /etc/puppet or whatever confdir is.

File: .pre-commit-config.yaml (hunk 5 above)
Comment (rhoml) [inline .pre-commit-config.yaml]: This is a file to use pre-commit.com at development time, pretty useful for githooks management

File: Gemfile (hunk 15 above)
Comment (rhoml) [inline Gemfile:3]: I'll fix this

File: Gemfile (hunk 18 above)
Comment (rhoml) [inline Gemfile]: I needed to test on our puppet version and forgot to roll it back

File: manifests/init.pp (hunk 20 above)
Comment (rhoml) [inline manifests/init.pp:24]: I didn't meant to change the location just wanted the config file to be related to the config dir but i missed the it was using the class settings. Can we make it a local variable instead?

Comment (rhoml) [inline manifests/init.pp:24]: Config_dir is pointing to $::settings::confdir so it didn't make sense that the config_path was different right?

File: .sync.yml (hunk 14 above)
Comment (rhoml) [inline .sync.yml]: https://github.com/puppetlabs/modulesync_configs#defining-module-files

File: manifests/init.pp (hunk 20 above)
Comment (rhoml) [inline manifests/init.pp:24]: I made a change to make `$config_path = "${config_dir}/config.yaml"` I think it makes better sense now right? and the same applies for the public and private keys which should depend on the `keys_dir` variable

File: .sync.yml (hunk 14 above)
Comment (ghoneycutt) [inline .sync.yml]: You use ModuleSync? This is for Puppet Labs' modules to sync updates to all of their modules.

**Comment by rhoml** [comment]
Removes sync.yml

**Comment by rhoml** [comment]
Hey @ghoneycutt any new on this? think it can be merged now?

**Comment by rhoml** [comment]
ping @ghoneycutt any news?

**Comment by rhoml** [comment]
Hey @ghoneycutt can we get this PR merged?

**Comment by rhoml** [comment]
@ghoneycutt is there any interest on merging this PR? otherwise I will have to fork the module

File: .rspec [hunk 21]
```diff
@@ -0,0 +1,4 @@
+--color
```
Comment (ghoneycutt) [inline .rspec]: This is not needed.

File: Gemfile [hunk 22]
```diff
@@ -6,10 +6,16 @@ else
...
 
+gem 'mocha'
+gem 'diff-lcs'
+gem 'json_pure'
+gem 'json'
 gem 'metadata-json-lint'
 gem 'puppetlabs_spec_helper', '>= 0.1.0'
 gem 'facter', '>= 1.7.0'
 gem 'rspec-puppet'
+gem 'rake', '~> 10.5.0'
```
Comment (ghoneycutt) [inline Gemfile]: Rake should be down in the ruby 1.8.7 conditional such as https://github.com/ghoneycutt/puppet-module-ssh/blob/master/Gemfile

File: Gemfile [hunk 23]
```diff
@@ -6,10 +6,16 @@ else
   gem 'puppet', :require => false
 end
 
+gem 'mocha'
```
Comment (ghoneycutt) [inline Gemfile]: These four are not needed

File: Rakefile [hunk 24]
```diff
@@ -1,5 +1,8 @@
 require 'puppetlabs_spec_helper/rake_tasks'
 require 'puppet-lint/tasks/puppet-lint'
+
+PuppetLint.configuration.fail_on_warnings = true
```
Comment (ghoneycutt) [inline Rakefile]: not sure these are actually needed

File: manifests/init.pp [hunk 25]
```diff
@@ -2,17 +2,124 @@
 #
 # Module to manage eyaml
 #
+# Variables
```
Comment (ghoneycutt) [inline manifests/init.pp]: Appreciated, but I document all of the parameters in the README. Check out https://github.com/ghoneycutt/puppet-module-ssh

File: manifests/init.pp [hunk 26]
```diff
@@ -2,17 +2,124 @@
...
+  $keys_dir            = '/etc/puppet/keys',
   $keys_dir_ensure     = 'directory',
-  $keys_dir_owner      = $::settings::user,
-  $keys_dir_group      = $::settings::group,
+  $keys_dir_owner      = root,
+  $keys_dir_group      = root,
   $keys_dir_mode       = '0500',
-  $public_key_path     = '/etc/puppet/keys/public_key.pkcs7.pem',
-  $private_key_path    = '/etc/puppet/keys/private_key.pkcs7.pem',
+  $public_key_path     = "${keys_dir}/public_key.pkcs7.pem",
```
Comment (ghoneycutt) [inline manifests/init.pp]: keep the hard coding of /etc/puppet/keys and Puppet now has issue with params using a previous param as a variable.

File: manifests/init.pp [hunk 27]
```diff
@@ -2,17 +2,124 @@
...
+# * `manage_keys`: Manage eyaml keys creation
+#    type: bool
+#    Default: undef
+#
 class eyaml (
   $package_name        = 'hiera-eyaml',
   $package_provider    = 'gem',
   $package_ensure      = 'present',
-  $keys_dir            = "${::settings::confdir}/keys",
+  $keys_dir            = '/etc/puppet/keys',
```
Comment (ghoneycutt) [inline manifests/init.pp:9]: good catch

File: manifests/init.pp (hunk 20 above)
Comment (ghoneycutt) [inline manifests/init.pp:24]: config_path should stay the same because of the params accessing previous params as variables that I mentioned.

**Comment by ghoneycutt** [comment]
I'm getting this ready for v1.0.0 and will be incorporating your changes. Thank you!

**Comment by rhoml** [comment]
@ghoneycutt changes are updated now and rebased

**Comment by rhoml** [comment]
Closing this PR as it is too far behind and I think you already implemented all the things proposed here anyways
